src/tools/rbd_mirror/ImageMap.cc

```diff
@@ -0,0 +1,663 @@
+     target_instance_id, on_map, on_unmap](int r) {
+      assert(m_lock.is_locked());
+
+      if (r < 0) {
+        derr << ": failed to shuffle global_image_id=" << global_image_id
+             << " from instance_id=" << source_instance_id << " to instance_id="
+             << target_instance_id << dendl;
+      }
+
+      set_retry_instance_id(global_image_id, Policy<I>::UNMAPPED_INSTANCE_ID);
+
+      delete on_map;
+      delete on_unmap;
+    });
+
+  ImageAction::Action action = ImageAction::Action::create_shuffle_action(
+    on_map, on_unmap, on_finish);
+  append_and_schedule_action(global_image_id, action);
+}
+
+template <typename I>
+void ImageMap<I>::update_images_added(const std::string &mirror_uuid,
+                                      const std::set<std::string> &global_image_ids) {
+  dout(20) << dendl;
+  assert(m_lock.is_locked());
+
+  for (auto const &global_image_id : global_image_ids) {
+    bool schedule_update = false;
+
+    std::string instance_id = m_policy->lookup(global_image_id);
+    if (instance_id == Policy<I>::UNMAPPED_INSTANCE_ID) {
+      instance_id = m_policy->map(global_image_id);
+
+      schedule_update = true;
+      m_image_action.init_state(global_image_id);
+      m_image_info.emplace(global_image_id, Info());
+    }
+
+    add_peer(global_image_id, mirror_uuid);
+    if (schedule_update) {
+      schedule_add_action(global_image_id, instance_id);
+    }
+  }
+}
+
+template <typename I>
+void ImageMap<I>::update_images_removed(const std::string &mirror_uuid,
+                                        const std::set<std::string> &global_image_ids) {
+  dout(20) << dendl;
+  assert(m_lock.is_locked());
+
+  Updates to_remove;
+
+  for (auto const &global_image_id : global_image_ids) {
+    std::string instance_id = m_policy->lookup(global_image_id);
+    assert(instance_id != Policy<I>::UNMAPPED_INSTANCE_ID);
+
+    if (!mirror_uuid.empty()) {
+      to_remove.emplace_back(global_image_id, instance_id);
+    }
+
+    bool schedule_update = remove_peer(global_image_id, mirror_uuid);
+    if (schedule_update) {
+      m_policy->unmap(global_image_id);
+      schedule_remove_action(global_image_id, instance_id);
+    }
+  }
+
+  // removal notification will be notified instantly. this is safe
+  // even after scheduling action for images as we still hold m_lock
+  notify_listener_remove_images(mirror_uuid, to_remove);
+}
+
+template <typename I>
+void ImageMap<I>::update_instances(const std::vector<std::string> &instances) {
+  dout(20) << dendl;
+
+  // global_image_id -> <from_instance, to_instance>
+  std::map<std::string, std::pair<std::string, std::string> > remap_ids;
+
+  {
+    Mutex::Locker locker(m_lock);
+    if (m_shutting_down) {
+      return;
+    }
+
+    m_policy->shuffle(instances, &remap_ids);
+
+    // track dead instances
+    for (auto const &instance : instances) {
+      if (!m_policy->instance_present(instance)) {
+        m_dead_instances.insert(instance);
```

Comment:
I am unclear about the purpose of ```m_dead_instances```  --- it seems that it's only tracking instances that don't have images mapped to them?